### PR TITLE
Implement timeout and ontimeout on XMLHttpRequest

### LINKED ISSUE #4648: No timeout on XMLHttpRequest
It seems that React Native is not implementing `timeout` [property](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/timeout) on XMLHttpRequest.

### PR DESCRIPTION
**Pull request by klvs**
Closes #4648

### CODE REVIEW

File: Libraries/Network/XMLHttpRequestBase.js [hunk 1]
```diff
@@ -199,6 +212,28 @@ class XMLHttpRequestBase {
...
+  }
+
+  // called by setTimeout, aborts the request
+  _checkTimeout(): void {
+    if (this.readyState !== this.DONE) {
+      // timed-out
+      this.abort();
+      if (this.ontimeout != null) {
+        // should send event, but not proccessing event, leaving empty
+        this.ontimeout(null);
```
Comment (satya164) [inline Libraries/Network/XMLHttpRequestBase.js]: Why pass null?

File: Libraries/Network/XMLHttpRequestBase.js [hunk 2]
```diff
@@ -199,6 +212,28 @@ class XMLHttpRequestBase {
...
+      }
+    }
+  }
+
+  // if a timeout is set, cancel the timeout call
+  _clearTimeout(): void {
+    if (this._timeoutId != null) {
+      clearTimeout(this._timeoutId);
```
Comment (satya164) [inline Libraries/Network/XMLHttpRequestBase.js]: Could be simplified to `this._timeoutId && clearTimeout(this._timeoutId);`

File: Libraries/Network/XMLHttpRequestBase.js [hunk 3]
```diff
@@ -199,6 +212,28 @@ class XMLHttpRequestBase {
...
+      this._timeoutId =  setTimeout(this._checkTimeout, this.timeout);
+    }
+  }
+
+  // called by setTimeout, aborts the request
+  _checkTimeout(): void {
+    if (this.readyState !== this.DONE) {
+      // timed-out
+      this.abort();
+      if (this.ontimeout != null) {
```
Comment (satya164) [inline Libraries/Network/XMLHttpRequestBase.js]: Could be simplified to `this.ontimeout && this.ontimeout();`

**Comment by nicklockwood** [comment]
I like that this is a pure-JS solution (because it works x-platform), but there is already a timeout set on the native side (60s by default on iOS: https://developer.apple.com/library/ios/documentation/Cocoa/Reference/Foundation/Classes/NSURLRequest_Class/#//apple_ref/occ/clm/NSURLRequest/requestWithURL:) so we need to find a way to integrate with that, otherwise setting a timeout higher than the native value on the JS side won't work.

**Comment by klvs** [comment]
@nicklockwood Thanks for reviewing.

It seems that the [specification for XMLHttpRequest](https://msdn.microsoft.com/en-us/library/cc304105(v=vs.85).aspx), if not given a timeout, is to wait indefinitely. That being said, we could just set the native request to never timeout.

However, after running a couple tests in chrome, it seems that XMLHttpRequest *will* timeout after 4 minutes flat and not wait indefinitely (I can provide code if anyone want to test this themselves). Reading a little bit deeper, from that same page:

````
If you set an XMLHttpRequest time-out value that is larger than the network stack's time-out value, 
the network stack will time out first and the ontimeout event will not be raised.
````

This is sort of the current behavior with our XMLHttpRequest in that it times out after 60 seconds. So the question is should an unspecified timeout value never time out, or should it have a reasonable default like chrome?

What are your thoughts on this? I haven't tested it yet, but does anyone know the default native timeout for android?

**Comment by nicklockwood** [comment]
I think it's reasonable to have a default and/or maximum timeout period rather than waiting forever if no timeout is specified, but I also think that the native stack timing out should appear the same way as a JS-set timeout.

With the current implementation, the native stack timing out will appear as an error, and won't fire the ontimeout handler.

If you'd prefer not to mess with the native code in this PR though, we could just implement the JS side to always set the timeout to <= 60s, then you should always get a timeout instead of an error.

We should also probably verify that the native default timeout on Android is the same.

**Comment by ide** [comment]
Doing this natively seems right to me since part of the reason of setting a timeout is to release system resources (minimal in this case). Imagine a polling program that issues an XHR with a 1-second timeout and then issues a new XHR after the previous XHR times out -- after a minute the native networking stack would still have 60 open connections.

**Comment by nicklockwood** [comment]
@ide well, it does call abort() after the timeout, so the native connection should still be cancelled after the JS side times out.

But I agree with you that it feels like timeout should be a native function, since it's already implemented on the native side anyway.

**Comment by ide** [comment]
Good point about `abort()`

**Comment by klvs** [comment]
@nicklockwood Right now it doesn't actually throw an error if the native stack times out but instead we just get request status of 0 and it doesn't trigger ontimeout (I agree this doesn't seem right, but it's what the XMLHttpRequest specification from msdn says). 

Given that I would prefer not to mess with native code in this PR (I can, but would rather do so in the future to ensure it's high quality), I need to set the default timeout (in JS) to <= 60s and then verify that android behaves the same way in order to get this PR accepted? Or is the consensus that it should be implemented natively and this way is not the right way?

**Comment by philikon** [comment]
> I need to set the default timeout (in JS) to <= 60s and then verify that android behaves the same way in order to get this PR accepted? Or is the consensus that it should be implemented natively and this way is not the right way? 

What this PR provides right now is implementable by anybody in JS already. If we're going to support the `timeout` and `ontimeout` properties, I for one would prefer if we'd hand off the timeout handling entirely to native. Disallowing 0 (to disable) or values over 60s seems half-hearted. Plus, I don't know what the default behaviour of OkHttp is, but I could imagine it's different from `NSURLConnection`.

**Comment by philikon** [comment]
In any case, please don't feel discouraged to submit a PR containing native code just because you might think it wouldn't meet the quality bar. We're happy to help. Also, you don't have to solve this problem for both platforms at the same time.

**Comment by klvs** [comment]
I think I've solved it for both platforms natively. Should I close this PR and submit it as a separate one or force push it here?

**Comment by nicklockwood** [comment]
@klvs if you can update this one then it would be nice to keep the discussion and history. But if it's easier to create a new PR you can always refer back to this, so I think it's fine either way.

File: Libraries/Network/XMLHttpRequest.android.js [hunk 4]
```diff
@@ -26,6 +26,12 @@ function convertHeadersMapToArray(headers: Object): Array<Header> {
 }
 
 class XMLHttpRequest extends XMLHttpRequestBase {
+  constructor() {
+    super();
+    // Native stack times out at 120s
+    // this.timeout = 120000;
+    this.timeout = 200000;
```
Comment (philikon) [inline Libraries/Network/XMLHttpRequest.android.js]: So why 200 secs instead of 120 then?

Comment (klvs) [inline Libraries/Network/XMLHttpRequest.android.js]: My mistake here. It was mistakenly left from testing. I'll fix it if it's being considered for a merge, but I'm about to push the native implementation. Sorry for the sloppy code.

**Comment by mkonicek** [comment]
@klvs Do you want to address the comments above and rebase this to make sure tests pass?

**Comment by klvs** [comment]
This is actually outdated and should be closed. It was addressed in #5038. Sorry for the straggling PR.

**Pull request closed without merging**